README.md
# * お し な が き *
MPD、License URLとヘッダ、PSSHをサーバーに送信するChrome用拡張機能。そしてデータを受信するためのローカルサーバーです。GYAO!でのみ動作します。


## 更新履歴
[最終更新: 2023-03-11 13:30](/CHANGELOG.md)

## なにができますか？
[こういうこと](https://ssbsblg.blogspot.com/2023/03/go-ahead.html)です。


## 動作環境
真価を発揮させるには WKS-KEYS が必要です。


### 参考
[WKS-KEYS｜初心者用](https://ssbsblg.blogspot.com/2023/03/wks-keys-for-beginners.html)  
[「無」からCDMを取り出す](https://ssbsblg.blogspot.com/2023/03/make-cdm-from-null.html)  


## インストール
- extensionフォルダを[拡張機能ページ](chrome://extensions/)にドロップします。  
（拡張機能のデベロッパーモードが有効になっている必要があります）
- server.py、main.bat を l3.py と同じディレクトリに配置します。


## 必要なこと
1) このコードを l3_mod.py として保存します。  
（場所は l3.py と同じ）

```python
import requests
from pywidevine.L3.cdm import deviceconfig
from base64 import b64encode
from pywidevine.L3.decrypt.wvdecryptcustom import WvDecrypt

def WV_Function(headers, pssh, lic_url):
	wvdecrypt = WvDecrypt(init_data_b64=pssh, cert_data_b64=None, device=deviceconfig.device_android_generic)
	widevine_license = requests.post(url=lic_url, data=wvdecrypt.get_challenge(), headers=headers)
	license_b64 = b64encode(widevine_license.content)
	wvdecrypt.update_license(license_b64)
	Correct, keyswvdecrypt = wvdecrypt.start_process()
	if Correct:
		return keyswvdecrypt
```

2) お使いの環境に合わせて main.bat を編集します。

### 参考
[command.txt](https://ssbsblg.blogspot.com/2023/03/command-txt.html)


## 使い方
1) サーバーを起動する。
2) 動画ページを開く。
3) 青色が点灯してから拡張機能ボタンを押す。


## 補足
- ひとたびインジケーターが青色になれば、タブをリロードしない限りデータは保持されたままです。  
サーバーの起動し忘れなどで送信失敗したとしても、またボタンを押せば同じデータが送信されます。  
(誤操作防止のため、ボタンには2秒間のタイムアウトを設定しています)

- main.bat は server.py の子プロセスとして実行されます。  
途中で server.py を閉じると、処理中のバッチファイルも閉じられます。

- 8000番ポートを使用します。  
もしも使用中の場合は変更することも可能です。  
server.py の最終行、background.js の33行目に該当の数値があります。


## 作った人
- [@AhhZahhMath (aka Sasabee)](https://note.com/sasabee)

noteから金銭的なサポートをしていただけると、とても励みになります。  
[noteの記事](https://note.com/sasabee/n/n6557be19006f)の下部にサポートボタンがありますので、よろしくお願いします。


CHANGELOG.md

# 更新履歴
このリポジトリ全体の主要な更新履歴です。


## 2023-03-11

### background.js
- ボタンの誤操作を防止する処理が複数タブに対応できていなかったので修正しました。
- ボタンクリック無効のタイムアウトを5秒から2秒に変更しました。


## 2023-03-10

### server.py
- 修正したはずのバッチファイルの名前が .bat になっていたので、今度こそ正しい main.bat に修正しました。


## 2023-03-09

### server.py
- 実行するバッチファイルの名前が sample.bat になっていたので、正しい main.bat に修正しました。

### README.md

- ポート番号の該当行を示す案内が間違っていたので修正しました。


extension/background.js
(()=>{
	let through = [];
	let buttonEnabled = [];
	const colors = {green:'#05E35E',red:'#E2041B',blue:'#058AE3',gray:'#808080',black:'#000000'};
	const buttonState = (bool, tabId)=>{
		bool ? buttonEnabled.push(tabId) : (buttonEnabled = buttonEnabled.filter(v=>v !== tabId));
	};
	const timer = (ms, tabId)=>{
		chrome.alarms.create(tabId + '', {when: Date.now() + ms});
	};
	const joinHeaders = (headers)=>{
		headers = headers.filter(v=>!v.name.includes('sec-ch-ua'))
		.map(v=>`"${v.name}":"${v.value}"`)
		.join(',\n');
		return `{${headers}}`;
	};
	const getPssh = async(tabId, dash)=>{
		await fetch(decodeURIComponent(dash))
		.then(response=>response.text())
		.then(text=>/(?<=widevine.+?pssh>).+?(?=<)/s.exec(text)[0])
		.then(async(v)=>{
			await chrome.storage.local.set({[`${tabId}__pssh`]:v});
		});
	};
	const postData = async(tabId)=>{
		let success = false;
		const obj = await chrome.storage.local.get([`${tabId}__headers`,`${tabId}__cenc`,`${tabId}__pssh`,`${tabId}__dash`,`${tabId}__title`]);
		const arr = Object.entries(obj);
		const data = {};
		for(let i = 0, l = arr.length; i <l; i++) {
			data[arr[i][0].split('__')[1]] = arr[i][1];
		}
		await fetch('http://localhost:8000/',{
			method:'POST',
			mode:'cors',
			headers:{
				'Content-Type':'text/plain'
			},
			body:JSON.stringify(data)
		})
		.then(()=>{
			success = true;
		})
		.catch(e=>{})
		.finally(()=>{
			chrome.action.setBadgeBackgroundColor({color:colors[success?'green':'red'],tabId:tabId});
		});
	};
	const sanitize = (str)=>{
		const before = ['\\','/',':','*','?','\"','<','>','|','　'];
		const after = ['￥','／','：','＊','？','”','＜','＞','｜',' '];
			for(let i = 0, l = before.length; i < l; i++) {
				str = str.split(before[i]).join(after[i]);
			}
		return str;
	};
	const storageOnChangedCallback = (changes, areaName)=>{
		const arr = Object.entries(changes);
		const tabId = arr[0][0].split('__')[0] - 0;
		if(!through.includes(tabId) && 'newValue' in arr[0][1]) {
			console.log(changes);
			(async()=>{
				const obj = await chrome.storage.local.get([`${tabId}__headers`,`${tabId}__cenc`,`${tabId}__pssh`,`${tabId}__dash`,`${tabId}__title`]);
				if(arr[0][0].includes('__pssh')) {
					through.push(tabId);
					buttonState(true, tabId);
					chrome.action.setBadgeBackgroundColor({color:colors.blue,tabId:tabId});
					console.log(`[Finish] ${obj[`${tabId}__title`]}`);
					return;
				}
				if(arr[0][0].includes('__headers') && Object.entries(obj).length === 3) {
					await chrome.scripting.executeScript({
						target: {tabId:tabId},
						func: ()=>document.querySelector('h1').innerText
					})
					.then(async(v)=>{
						await chrome.storage.local.set({[`${tabId}__title`]:sanitize(v[0].result)});
					});
					await getPssh(tabId,obj[`${tabId}__dash`]);
				}
			})();
		}
	};
	const tabsOnUpdatedCallback = (tabId, changeInfo, tab)=>{
		if(['https://gyao.yahoo.co.jp/episode/','https://gyao.yahoo.co.jp/title/'].some(v=>tab.url.includes(v))) {
			if(changeInfo.status === 'loading') {
				(async()=>{
					const arr = [`${tabId}__headers`,`${tabId}__cenc`,`${tabId}__pssh`,`${tabId}__dash`,`${tabId}__title`];
					for(let i = 0, l = arr.length; i < l; i++) {
						await chrome.storage.local.remove(arr[i]);
					}
					through = through.filter(v=>v !== tabId);
					buttonEnabled = buttonEnabled.filter(v=>v !== tabId);
					buttonState(false, tabId);
					chrome.action.setBadgeText({text:' ',tabId:tabId});
					if(!tab.title.includes('ページが見つかりませんでした')) {
						console.log(`[Start] ${tab.title}`);
						chrome.action.setBadgeBackgroundColor({color:colors.gray,tabId:tabId});
					} else {
						console.log('[404] Not Found');
						through.push(tabId);
						chrome.action.setBadgeBackgroundColor({color:colors.black,tabId:tabId});
					}
				})();
			}
		}
	};
	const tabsOnRemovedCallback = (tabId, removeInfo)=>{
		(async()=>{
			through = through.filter(v=>v !== tabId);
			const arr = [`${tabId}__headers`,`${tabId}__cenc`,`${tabId}__pssh`,`${tabId}__dash`,`${tabId}__title`];
			for(let i = 0, l = arr.length; i < l; i++) {
				await chrome.storage.local.remove(arr[i]);
			}
		})();
	};
	const onBeforeRequestFilter = {urls: ['https://manifest.prod.boltdns.net/manifest/*','https://manifest.prod.boltdns.net/license/*']};
	const onBeforeRequestCallback = (details)=>{
			const tabId = details.tabId;
			if(!through.includes(tabId) && !details.frameId) {
				if(details.url.includes('/hls/')) {
					through.push(tabId);
					chrome.action.setBadgeBackgroundColor({color:colors.black,tabId:tabId});
					console.log(`[Stop] Is Not DRM`);
					return;
				}
				(async()=>{
					const arr = ['dash','cenc'];
					const obj = await chrome.storage.local.get([`${tabId}__dash`,`${tabId}__cenc`]);
					for(let i = 0; i < 2; i++) {
						if(details.url.includes(`/${arr[i]}/`) && !(`__${arr[i]}` in obj)) {
							await chrome.storage.local.set({[`${tabId}__${arr[i]}`]:details.url});
							return;
						}
					}
				})();
			}
	};
	const onBeforeSendHeadersFilter = {urls: ['https://manifest.prod.boltdns.net/license/*']};
	const onBeforeSendHeadersCallback = (details)=>{
		const tabId = details.tabId;
		if(!through.includes(tabId)) {
			(async()=>{
				const obj = await chrome.storage.local.get(`${tabId}__headers`);
				if(!(`__headers` in obj)) {
					const headers = joinHeaders(details.requestHeaders);
					await chrome.storage.local.set({[`${tabId}__headers`]:headers});
				}
			})();
		}
	};
	const alarmsOnAlarmCallback = (alarm)=>{
		buttonState(true, alarm.name - 0);
	};
	const actionOnClickedCallback = (tab)=>{
		buttonEnabled.includes(tab.id) && (async()=>{
			buttonState(false, tab.id);
			await postData(tab.id);
			timer(2000, tab.id);
		})();
	};
	const runtimeOnCallback = ()=>{
		(async()=>{
			await chrome.storage.local.clear();
			console.log('[Init] Remove All Data');
		})();
	};
	chrome.action.onClicked.addListener(actionOnClickedCallback);
	chrome.alarms.onAlarm.addListener(alarmsOnAlarmCallback);
	chrome.runtime.onInstalled.addListener(runtimeOnCallback);
	chrome.runtime.onStartup.addListener(runtimeOnCallback);
	chrome.storage.onChanged.addListener(storageOnChangedCallback);
	chrome.tabs.onRemoved.addListener(tabsOnRemovedCallback);
	chrome.tabs.onUpdated.addListener(tabsOnUpdatedCallback);
	chrome.webRequest.onBeforeRequest.addListener(onBeforeRequestCallback,onBeforeRequestFilter,['requestBody','extraHeaders']);
	chrome.webRequest.onBeforeSendHeaders.addListener(onBeforeSendHeadersCallback,onBeforeSendHeadersFilter,['requestHeaders','extraHeaders']);
})();


server.py
from wsgiref.simple_server import make_server
from time import time_ns
from subprocess import Popen
from json import dumps
from json import loads
from l3_mod import WV_Function as f

def app(env, start_response):
	if env.get('REQUEST_METHOD') == 'POST':
		return post(env, start_response)
	else:
		data = dumps({'message': '400 Bad Request'})
		start_response('404 Not Found', [('Content-type', 'application/json; charset=utf-8')])
		return [bytes(data, 'utf8')]

def post(env, start_response):
	content_length = env.get('CONTENT_LENGTH', 0)
	body = env.get('wsgi.input').read(int(content_length)).decode('utf8')
	d = loads(body)
	keys = f(loads(d['headers']), d['pssh'], d['cenc'])
	print(keys)
	l = d['dash'].split('=')
	mpd = l[0]
	token = l[1]
	Popen(['main.bat', d['title'], keys[0], mpd, token, f'{time_ns()}'[:11]])
	start_response('200 OK', [
		('Content-type', 'text/plain; charset=utf-8'),
		('Access-Control-Allow-Origin', env.get('HTTP_ORIGIN'))
	])
	return []

if __name__ == '__main__':
	make_server('localhost', 8000, app).serve_forever()
